### PR TITLE
Integration of Redis Tool for Efficient Token Caching

### LINKED ISSUE #8153: Integration of Redis Tool for Efficient Token Caching
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

Yes, I frequently need to request tokens for third-party API calls. This process can be time-consuming and inefficient.
I would like to see a Redis tool integrated into Dify. This tool would allow users to cache tokens, reducing the need for repeated token requests.
By caching tokens, I can save time and resources, making my workflow more efficient. This feature would eliminate the need to request a new token for each API call, thereby speeding up the development process.

### 2. Additional context or comments

I have developed a function in my local code that implements this feature. I am willing to share this function with the Dify community.

### 3. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by G81192**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description
I have developed a function in my local code that implements this feature. I am willing to share this function with the Dify community. for issue #8153

- I frequently need to request tokens for third-party API calls. This process can be time-consuming and inefficient.
- I would like to see a Redis tool integrated into Dify. This tool would allow users to cache tokens, reducing the need for repeated token requests.
- By caching tokens, I can save time and resources, making my workflow more efficient. This feature would eliminate the need to request a new token for each API call, thereby speeding up the development process.

Close https://github.com/langgenius/dify/issues/8153

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      pt_BR: value
+    human_description:
+      en_US: value
+      zh_Hans: value
+      pt_BR: value
+    llm_description: value
+    form: llm
+  - name: expire
+    type: number
+    required: true
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml:42]: Better to have default value for expiration time.

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      en_US: value
+      zh_Hans: value
+      pt_BR: value
+    human_description:
+      en_US: value
+      zh_Hans: value
+      pt_BR: value
+    llm_description: value
+    form: llm
+  - name: expire
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]:
```suggestion
  - name: expiration
```

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    type: number
+    required: true
+    label:
+      en_US: expire(s)
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]:
```suggestion
      en_US: expire time in second
```

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      zh_Hans: 过期时间(秒)
+      pt_BR: expire(s)
+    human_description:
+      en_US: expire(s)
+      zh_Hans: 过期时间(秒)
+      pt_BR: expire(s)
+    llm_description: expire(s)
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]: ditto for explanation of time unit for human and LLM

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    form: llm
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]: Should not be decided by LLM.

File: api/core/tools/provider/builtin/redis_tool/tools/decorator.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        kwargs['key'] = KEY_PREFIX + kwargs.get('key', '')
+
+        logging.info(f"[{host}:{port}]Redis connection.")
+        redis_client = redis.StrictRedis(
+            host=host,
+            port=port,
+            db=db,
+            password=password,
+            decode_responses=True
+        )
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/decorator.py]: To support timeout as well by default.

**Review by bowenliang123 — COMMENTED** [review]
Some comments to follow.

**Comment by G81192** [comment]
hi ,fix all the comments

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    label:
+      en_US: key
+      zh_Hans: key
+      pt_BR: key
+    human_description:
+      en_US: key
+      zh_Hans: key
+      pt_BR: key
+    llm_description: key
+    form: llm
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]:
```suggestion
    form: form
```
The value of `key` should not be decided by LLM, which is very dangerous.

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    label:
+      en_US: value
+      zh_Hans: value
+      pt_BR: value
+    human_description:
+      en_US: value
+      zh_Hans: value
+      pt_BR: value
+    llm_description: value
+    form: llm
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]: How could the LLM determine the proper value data type from the LLM description anyway?

File: api/core/tools/provider/builtin/redis_tool/tools/get_string.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    required: true
+    label:
+      en_US: key
+      zh_Hans: key
+      pt_BR: key
+    human_description:
+      en_US: key
+      zh_Hans: key
+      pt_BR: key
+    llm_description: key
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/redis_tool/tools/get_string.yaml]: All the descriptions for human and LLM is unreadable and not helpful for generation.

**Review by bowenliang123 — CHANGES_REQUESTED** [review]
Please properly decide the descriptions of the tool parameters for both human and LLM, which is the key point for usage and runtime.

**Comment by bowenliang123** [comment]
cc @laipz8200 @JohnJyong

**Comment by hjlarry** [comment]
I am puzzled is it insecure to have users fill in the Redis connection details ?
for example,  can I use a script to continuously request to the server like `cloud.dify.ai` to guess the password ?

File: api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml (hunk 7 above)
Comment (G81192) [reply api/core/tools/provider/builtin/redis_tool/tools/set_string.yaml]: you're right

**Comment by G81192** [comment]
@hjlarry The `secret-input` type is encrypted, and all secret use this mode. I don't think this is a very serious issue, as long as you manage permissions properly.

**Comment by hjlarry** [comment]
> @hjlarry The `secret-input` type is encrypted, and all secret use this mode. I don't think this is a very serious issue, as long as you manage permissions properly.

the encrypted password  can defense user's input can't be stolen.  but if I don't know the redis password,  can I brute force attack your platform?

another issue is the api-key of openAI/gemini is encrypted store in the database, but not encrypted in the redis,  everyone who can use this tool  seems can read your api-key.

**Comment by G81192** [comment]
> > @hjlarry The `secret-input` type is encrypted, and all secret use this mode. I don't think this is a very serious issue, as long as you manage permissions properly.
> 
> the encrypted password can defense user's input can't be stolen. but if I don't know the redis password, can I brute force attack your platform?
> 
> another issue is the api-key of openAI/gemini is encrypted store in the database, but not encrypted in the redis, everyone who can use this tool seems can read your api-key.


Regarding your two points, my thoughts :
1. The Redis tool is used as a node within the process and can only be accessed during workflow editing. I'm not entirely sure how the brute force attack you mentioned operates, so I'm uncertain about that.
2. If within the same workspace, it is indeed usable by all. However, I believe that being in the same workspace implies trust, and similarly, all LLM tokens can be used within the same workspace.

**Comment by hjlarry** [comment]
1. the brute force attack maybe like this way.  I can continuous change the api_key and send request until the response is OK
```
curl 'https://cloud.dify.ai/console/api/workspaces/current/tool-provider/builtin/perplexity/update' \
  -H 'content-type: application/json' \
  -H 'referer: https://cloud.dify.ai/tools' \
  --data-raw '{"credentials":{"perplexity_api_key":"123"}}'
```
I can't steal a perplexity_api_key because the key is too long and they have defense system.  But we can't makesure everyone use a strong redis password, right?

2. yeah, the same workspace's users can use the same LLM, but they can't view the LLM's api-key. 


I am not sure if these concerns make sense, just some thoughts.

**Comment by G81192** [comment]
> 1. the brute force attack maybe like this way.  I can continuous change the api_key and send request until the response is OK
> 
> ```
> curl 'https://cloud.dify.ai/console/api/workspaces/current/tool-provider/builtin/perplexity/update' \
>   -H 'content-type: application/json' \
>   -H 'referer: https://cloud.dify.ai/tools' \
>   --data-raw '{"credentials":{"perplexity_api_key":"123"}}'
> ```
> 
> I can't steal a perplexity_api_key because the key is too long and they have defense system. But we can't makesure everyone use a strong redis password, right?
> 
> 2. yeah, the same workspace's users can use the same LLM, but they can't view the LLM's api-key.
> 
> I am not sure if these concerns make sense, just some thoughts.

Your concern is valid, as security is a red line.
1. If this method can be cracked, then it could be cracked through direct connection as well. This is more of a weak password security issue rather than a tool problem.
2. With the `secret-input` type, Redis passwords cannot be viewed either.
This is just to cache some data used during the process to improve efficiency. Therefore, I also do not recommend sharing the business Redis instance and suggest maintaining proper data isolation

**Comment by hjlarry** [comment]
1. the redis-server / database-server  always only can be accessed by the api server, it's password always config in the `.env` file. The tool's authorzation make a bridge between redis and frontend user,  the user can direct try to connect to the redis-server
2. if it's a separated redis instance, it seems ok

**Comment by G81192** [comment]
> hi ,fix all the comments

@bowenliang123 
hi,Is there anything else in this PR that needs to be modified?

**Comment by bowenliang123** [comment]
Thanks for the feedback. I would suggest Dify team to decide whether they would accept this tool first before further detailed polishment.

**Comment by G81192** [comment]
> Thanks for the feedback. I would suggest Dify team to decide whether they would accept this tool first before further detailed polishment.

That's fine.

**Comment by crazywoola** [comment]
Sorry, we do not accept new tools for now, we plan to have a tool market place int he future. :)

**Pull request closed without merging**